lib/improver/argparser.py

```diff
@@ -28,19 +28,185 @@
 # CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 # ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 # POSSIBILITY OF SUCH DAMAGE.
-"""Common option utilities for Improver CLIs."""
+"""Common option utilities for improver CLIs."""
 
-import argparse
+from argparse import ArgumentParser
 
 
-class ArgParser(argparse.ArgumentParser):
-    """Argument parser for improver CLIs"""
+class ArgParser(ArgumentParser):
+    """Argument parser for improver CLIs.
+    (Subclass of argparse.ArgumentParser)"""
+
+    # CENTRALIZED_ARGUMENTS, is a dictionary where, for each element:
+    #   the key is a string representing the argument name,
+    #   the value is a list containing 2 elements:
+    #       1. a list of strings containing the different flags
+    #       2. a dictionary containing all of the kwargs which are passed
+    #          to the ArgumentParser.add_argument() method
+
+    # All CLIs should select something from this dictionary:
+    # NB: --help included for free with ArgumentParser
+    CENTRALIZED_ARGUMENTS = {
+        'input_file': (
+            ['input_filepath'],
+            {}),
+        'output_file': (
+            ['output_filepath'],
+            {}),
+    }
+
+    # *All* CLIs will use the options here (no option to disable them):
+    COMPULSORY_ARGUMENTS = {
+        # TODO: Implement (some of?) these:
+        #        'debug': (
+        #            ['--debug'],
+        #            {'action': 'store_true'}),
+        #        'dry_run': (
+        #            ['--dry-run'],
+        #            {'action': 'store_true'}),
+        #        'profile' : (
+        #            ['--profile'],
+        #            {'action': 'store_true'}),
+        #        'verbose': (
+        #            ['--verbose', '-v'],
+        #            {'action': 'count'}),
+        #        'version': (
+        #            ['--version'],
+        #            {'action': 'version',
+        #             'version': improver.version_message() }),
+    }
+
+    # We can override including these, but options common to everything should
+    # be in a list here:
+    # DEFAULT_CENTRALIZED_ARGUMENTS = ('input_file', 'output_file')
+    DEFAULT_CENTRALIZED_ARGUMENTS = ()
+
+    def __init__(self, central_arguments=DEFAULT_CENTRALIZED_ARGUMENTS,
+                 specific_arguments=None, **kwargs):
+        """Create an ArgParse instance, which is a subclass of
+        argparse.ArgumentParser and automatically add all of the arguments.
+        (Note: The ArgParse.COMPULSORY_ARGUMENTS are always added.)
+
+        Args:
+            central_arguments (list):
+              A list containing the centralized arguments we require.
+              (Keys of the centralized argument dictionary). By default this is
+              set as ArgParse.DEFAULT_CENTRALIZED_ARGUMENTS.
+            specfic_arguments (list):
```

Comment:
"specific_arguments"

Reply:
That typo has caught me out more times...